etcd/watchmanager/manager.go

```diff
@@ -0,0 +1,148 @@
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	scope := opts.InstrumentsOptions().MetricsScope()
+	return &manager{
+		opts:   opts,
+		logger: opts.InstrumentsOptions().Logger(),
+		m: metrics{
+			etcdWatchCreate: scope.Counter("etcd-watch-create"),
+			etcdWatchError:  scope.Counter("etcd-watch-error"),
+			etcdWatchReset:  scope.Counter("etcd-watch-reset"),
+		},
+		updateFn:      opts.UpdateFn(),
+		tickAndStopFn: opts.TickAndStopFn(),
+	}, nil
+}
+
+type manager struct {
+	opts   Options
+	logger xlog.Logger
+	m      metrics
+
+	updateFn      UpdateFn
+	tickAndStopFn TickAndStopFn
+}
+
+type metrics struct {
+	etcdWatchCreate tally.Counter
+	etcdWatchError  tally.Counter
+	etcdWatchReset  tally.Counter
+}
+
+func (w *manager) watchChanWithTimeout(key string) (clientv3.WatchChan, error) {
+	doneCh := make(chan struct{})
+
+	var watchChan clientv3.WatchChan
+	go func() {
+		watchChan = w.opts.Watcher().Watch(
+			clientv3.WithRequireLeader(context.Background()),
+			key,
+			w.opts.WatchOptions()...,
+		)
+		close(doneCh)
+	}()
+
+	select {
+	case <-doneCh:
+		return watchChan, nil
+	case <-time.After(w.opts.WatchChanInitTimeout()):
+		return nil, fmt.Errorf("etcd watch chan creation timed-out on key: %s", key)
+	}
+}
+
+func (w *manager) Watch(key string) {
+	ticker := time.Tick(w.opts.WatchChanCheckInterval())
+
+	var (
+		watchChan clientv3.WatchChan
+		err       error
+	)
+	for {
+		if watchChan == nil {
+			w.m.etcdWatchCreate.Inc(1)
+			watchChan, err = w.watchChanWithTimeout(key)
+			if err != nil {
+				w.logger.Errorf("could not create etcd watch, %v", err)
+
+				// NB(cw) when we failed to create a etcd watch channel
+				// we do a get for now and will try to recreate the watch chan later
+				if err = w.updateFn(key); err != nil {
+					w.logger.Errorf("failed to get value for key %s: %v", key, err)
+				}
+				// avoid recreating watch channel too frequently
+				time.Sleep(w.opts.WatchChanResetInterval())
+				continue
+			}
+		}
+
+		select {
+		case r, ok := <-watchChan:
+			if !ok {
+				// the watch chan is closed, set it to nil so it will be recreated
+				// this is unlikely to happen but just to be defensive
+				watchChan = nil
+				w.logger.Warnf("etcd watch channel closed on key %s, recreating a watch channel", key)
+
+				// avoid recreating watch channel too frequently
+				time.Sleep(w.opts.WatchChanResetInterval())
+				w.m.etcdWatchReset.Inc(1)
+
+				continue
+			}
+
+			// handle the update
+			if err = r.Err(); err != nil {
+				w.logger.Errorf("received error on watch channel: %v", err)
+				w.m.etcdWatchError.Inc(1)
+				// do not stop here, even though the update contains an error
+				// we still take this chance to attemp a Get() for the latest value
+			}
+
+			if err = w.updateFn(key); err != nil {
+				w.logger.Errorf("received notification for key %s, but failed to get value: %v", key, err)
+			}
+		case <-ticker:
+			if w.tickAndStopFn(key) {
```

Comment:
Instead of reading a ticker and check if it should stop, wouldn't it be better to immediately signal stop through some channel?  Otherwise we're going to be waking up goroutines that don't necessarily need to be woken.